Editor preview

caseyflynn-google opened the pull request:
Fix #1252 Add support for a preview code editor.

vince-fugnitto commented on packages/navigator/src/browser/navigator-preferences.ts:
```diff
@@ -34,13 +34,19 @@ Configure glob patterns for excluding files and folders from the navigator. A re
+        },
+        'navigator.singleClickPreview': {
+            type: 'boolean',
+            description: 'Single-click or selection of a file in the Navigator open the given file in preview mode. If no preview is available for the given file type, single click will open the file.',
```
minor typo, I think you meant `opens` rather than `open`.
perhaps the description can be shortened (it will most likely be cut in the preferences widget tree) to something more easily read by end-users.

example: ` controls whether editors are open as previews when selected or single-clicked` ? 
we can debate however on what the description can ultimately be.

vince-fugnitto commented on packages/editor-preview/src/browser/editor-preview-manager.ts:
```diff
@@ -0,0 +1,125 @@
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
```
minor, missing new line between end of copyright and import statements

vince-fugnitto commented on packages/editor-preview/src/browser/editor-preview-manager.ts:
```diff
@@ -0,0 +1,125 @@
+import { EditorManager,  EditorOpenerOptions, EditorWidget, } from '@theia/editor/lib/browser';
+import { EditorPreviewWidget } from './editor-preview-widget';
+import { EditorPreviewWidgetFactory, EditorPreviewWidgetOptions } from './editor-preview-factory';
+import { WidgetOpenHandler, WidgetOpenerOptions } from '@theia/core/lib/browser';
```
minor, new line missing between end of imports and jsdoc

vince-fugnitto commented on packages/editor-preview/src/browser/editor-preview-widget.ts:
```diff
@@ -0,0 +1,226 @@
+            this.pinListeners.push({dispose: () => tabBar.tabMoved.disconnect(tabMovedListener)});
+
+            requestAnimationFrame(() => {
+                const tabNode = tabBar.contentNode.children.item(tabBar.currentIndex);
```
`tabNode` can possibly be null but no null checks are performed. The error fails the compilation and cli.

vince-fugnitto commented:
@caseyflynn-google you are missing the `package.spec.ts` which can be found in other packages. 
This is used for the cli to ensure it passes.

You are also missing the `@theia/editor-preview` dependency in the `package.json` found under the example/electron folder.

vince-fugnitto commented on packages/navigator/src/browser/navigator-preferences.ts:
```diff
@@ -34,13 +34,19 @@ Configure glob patterns for excluding files and folders from the navigator. A re
 export interface FileNavigatorConfiguration {
     'navigator.autoReveal': boolean;
     'navigator.exclude': { [key: string]: boolean };
+    'navigator.singleClickPreview': boolean;
```
maybe the name `enablePreview` like vscode would be better ?

akosyakov commented:
@caseyflynn-google cool! really like that optional and not invasive

I've noticed following glitches:
- if I pin one editor and then preview another then switching between them by clicking on corresponding files in the navigator does not work
- if I start typing in a preview editor it gets closed and a normal editor gets opened, would it be possible to reuse preview editor?
- not only a caption of preview editor is in italic style, but an icon as well
- clicking on a file in the navigator activates the preview editor, it would be better if the focus stays on the navigator, otherwise i never can select a file in the navigator by clicking on it

akosyakov commented on packages/editor-preview/src/browser/editor-preview-widget.ts:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    getResourceUri(): URI | undefined {
+        return this.editorWidget_!.getResourceUri();
```
`this.editorWidget_ && this.editorWidget_.getResourceUri()`  is safer

akosyakov commented on packages/editor-preview/src/browser/editor-preview-widget.ts:
```diff
@@ -0,0 +1,226 @@
+
+    get saveable(): Saveable|undefined {
+        if (this.editorWidget_) {
+            return this.editorWidget_!.saveable;
```
`!` is redundant

akosyakov commented on packages/editor-preview/src/browser/editor-preview-widget.ts:
```diff
@@ -0,0 +1,226 @@
+        return this.editorWidget_!.createMoveToUri(resourceUri);
+    }
+
+    pinEditorWidget() {
```
please put return types explicitly, even when they are  `void` to avoid accidental breaking changes

akosyakov commented on packages/editor-preview/src/browser/editor-preview-widget.ts:
```diff
@@ -0,0 +1,226 @@
+        this.onDidChangeTrackableWidgetsEmitter.fire([w]);
+    }
+
+    onAfterAttach(msg: Message) {
```
`protected`

akosyakov commented on packages/editor-preview/src/browser/editor-preview-widget.ts:
```diff
@@ -0,0 +1,226 @@
+        this.addTabPinningLogic();
+    }
+
+    private lastParent: DockPanel;
```
`DockPanel | undefined`

akosyakov commented on packages/editor-preview/src/browser/editor-preview-widget.ts:
```diff
@@ -0,0 +1,226 @@
+
+    protected onResize(msg: Widget.ResizeMessage): void {
+        if (this.editorWidget_) {
+            if (msg.width < 0 || msg.height < 0) {
```
Have you tried just to forward a message to an editor?
```ts
MessageLoop.sendMessage(this.editorWidget_, msg);
```

caseyflynn-google replied:
I updated to pass the message, but we will still need to set the size until autoSizing is working. if you like we can move the code to the editor widget.

caseyflynn-google replied:
The way I read / interpreted this is the resize will not be performed here: https://github.com/theia-ide/theia/blob/e8dd128c25294b23741cc345067952ef7cd85864/packages/monaco/src/browser/monaco-editor.ts#L291 due to autoSizing being false.

akosyakov commented on packages/navigator/src/browser/navigator-preferences.ts:
```diff
@@ -34,13 +34,19 @@ Configure glob patterns for excluding files and folders from the navigator. A re
+        'navigator.singleClickPreview': {
+            type: 'boolean',
+            description: 'Single-click or selection of a file in the Navigator open the given file in preview mode. If no preview is available for the given file type, single click will open the file.',
+            default: false
```
How about `true` by default?

caseyflynn-google commented:
> @caseyflynn-google cool! really like that optional and not invasive
> 
> I've noticed following glitches:
> 
> * if I pin one editor and preview another then switching between them by clicking on corresponding files in the navigator does not work

Fixed.

> * if I start typing in a preview editor it gets closed and a normal editor gets opened, would it be possible to reuse the preview editor?

I believe this is the current behavior.

> * not only a caption of preview editor is in italic style, but an icon as well

Fixed.

> * clicking on a file in the navigator activates the preview editor, it would be better if the focus stays on the navigator, otherwise, i never can select a file in the navigator by clicking on it

I am not really sure how to accomplish this, any suggestions would be appreciated.

caseyflynn-google commented:
> @caseyflynn-google you are missing the `package.spec.ts` which can be found in other packages.
> This is used for the cli to ensure it passes.
> 
> You are also missing the `@theia/editor-preview` dependency in the `package.json` found under the example/electron folder.
> 
> I noticed you introduced a lot of code as part of your PR, but it would also be nice if you included some unit tests as to not to decrease Theia's overall test coverage.

Done.

akosyakov commented on packages/navigator/src/browser/navigator-model.ts:
```diff
@@ -132,6 +134,13 @@ export class FileNavigatorModel extends FileTreeModel {
 
+    selectNode(node: SelectableTreeNode): void {
+        if (FileNode.is(node) && this.preferences['navigator.enablePreview']) {
+            open(this.openerService, node.uri, {preview: true});
```
you should be able to pass `mode: ' reveal'` here to preserve focus

caseyflynn-google replied:
Ah! Thanks for this! To get consistent focus, I used {mode: reveal} in the widget manager and made a few other minor changes.

akosyakov replied:
We need reveal here also, otherwise if editor preview is disabled, then focus is again stolen from the navigator.

caseyflynn-google commented:
> > @caseyflynn-google cool! really like that optional and not invasive
> > I've noticed following glitches:
> > 
> > * if I pin one editor and preview another then switching between them by clicking on corresponding files in the navigator does not work
> 
> Fixed.
> 
> > * if I start typing in a preview editor it gets closed and a normal editor gets opened, would it be possible to reuse the preview editor?
> 
> I believe this is the current behavior.
> 
> > * not only a caption of preview editor is in italic style, but an icon as well
> 
> Fixed.
> 
> > * clicking on a file in the navigator activates the preview editor, it would be better if the focus stays on the navigator, otherwise, i never can select a file in the navigator by clicking on it
> 
> I am not really sure how to accomplish this, any suggestions would be appreciated.

Got this fixed.

svenefftinge commented:
> @marcdumais-work @svenefftinge Is it possible to add @caseyflynn-google as a Eclipse Theia committer?

Sure, would be great to have you Casey! Are you interested?

akosyakov commented on packages/navigator/src/browser/navigator-preferences.ts:
```diff
@@ -34,13 +34,19 @@ Configure glob patterns for excluding files and folders from the navigator. A re
             }
-        }
+        },
+        'navigator.enablePreview': {
```
It would be better if we split this preference to 2 preferences as in VS Code:
- `list.openMode` should be defined in core, possible values `singleClick` or `doubleClick`, `singleClick` is default. It should be used in different widgets, not only in the navigator, e.g. in git and git history widgets
- `editor.enablePreview` should be defined in `editor-preview` extension, of `boolean` type, `true` is default. If it is `false` the preview extension should not do anything.

See https://stackoverflow.com/questions/40336583/how-to-disable-preview-file-with-single-click-in-vs-code

akosyakov replied:
Also, git and history widgets should pass `preview` as true when they call to the opener service. It could be done as a separate PR.

caseyflynn-google replied:
Updated preferences. I will file a follow up PR for the git and history widgets after this has been submitted.

marcdumais-work commented:
> @marcdumais-work @svenefftinge Is it possible to add @caseyflynn-google as a Eclipse Theia committer?

Sure, sounds like a good idea, at least eventually. 

Someone would need to nominate him through the Eclipse Foundation tool for that effect, providing a solid case to justify the nomination, to the current committers' satisfaction.

Usually it means listing significant contributions to the project, and making the case of how they warrant the person being made a committer (e.g. some mix of quantity/quality/complexity/rarity of the documented contributions). I have not followed closely enough to know whether that's arguably the case, here, yet. 

More details: https://wiki.eclipse.org/Development_Resources/HOWTO/Nominating_and_Electing_a_New_Committer
https://www.eclipse.org/projects/handbook/#elections-nomination

marcdumais-work commented on packages/editor-preview/README.md:
```diff
@@ -0,0 +1,7 @@
+# Theia - Editor Preview Extension
+
```
It would be nice to have a description of what this extension brings. Maybe some bits from the overview part from this comment could be copied:  https://github.com/theia-ide/theia/issues/1252#issuecomment-428766722

caseyflynn-google replied:
https://github.com/theia-ide/theia/pull/3561

marcdumais-work commented:
@caseyflynn-google impressive (first I think?) contribution. Good analysis of the problem in #1252. I just tried the PR on Gitpod, and I think I'll use it daily once merged. I see that beyond the immediately apparent feature, you've added infrastructure to make it easier to support preview for other widgets - even nicer. Thanks! 

Just a heads-up: this PR is getting close to the 1000 LoC limit. Over that we will need to submit a CQ, which would delay a bit merging. It looks like it's almost ready to merge, in any case?

caseyflynn-google commented:
Absolutely.

On Fri, Nov 9, 2018, 12:43 AM Sven Efftinge <notifications@github.com wrote:

> @marcdumais-work <https://github.com/marcdumais-work> @svenefftinge
> <https://github.com/svenefftinge> Is it possible to add @caseyflynn-google
> <https://github.com/caseyflynn-google> as a Eclipse Theia committer?
>
> Sure, would be great to have you Casey! Are you interested?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theia-ide/theia/pull/3373#issuecomment-437289981>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AWYExHJACgkMtnj-9yKnj6_f5OzbMq7hks5utUBFgaJpZM4YKY-1>
> .
>

caseyflynn-google commented:
I think this is ready to merge, I will update the description and add
preview support to the guy and history widgets in a follow-up PR.

On Fri, Nov 9, 2018, 5:38 PM Marc Dumais <notifications@github.com wrote:

> @caseyflynn-google <https://github.com/caseyflynn-google> impressive
> (first I think?) contribution. Good analysis of the problem in #1252
> <https://github.com/theia-ide/theia/issues/1252>. I just tried the PR on
> Gitpod, and I think I'll use it daily once merged. I see that beyond the
> immediately apparent feature, you've added infrastructure to make it easier
> to support preview for other widgets - even nicer. Thanks!
>
> Just a heads-up: this PR is getting close to the 1000 LoC limit. Over that
> we will need to submit a CQ, which would delay a bit merging. It looks like
> it's almost ready to merge, in any case?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theia-ide/theia/pull/3373#issuecomment-437546775>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AWYExKWNhqGq1mf-RokfUKXRr9s1kA_cks5uti4ugaJpZM4YKY-1>
> .
>

caseyflynn-google commented:
> @caseyflynn-google impressive (first I think?) contribution. Good analysis of the problem in #1252. I just tried the PR on Gitpod, and I think I'll use it daily once merged. I see that beyond the immediately apparent feature, you've added infrastructure to make it easier to support preview for other widgets - even nicer. Thanks!
> 
Glad to hear this is a useful feature!

> Just a heads-up: this PR is getting close to the 1000 LoC limit. Over that we will need to submit a CQ, which would delay a bit merging. It looks like it's almost ready to merge, in any case?

I think this is ready to merge. I will add additional documentation in the README.md after this has merged to keep it under the 1000 LoC limit.

caseyflynn-google commented:
> > @marcdumais-work @svenefftinge Is it possible to add @caseyflynn-google as a Eclipse Theia committer?
> 
> Sure, would be great to have you Casey! Are you interested?

Absolutely.

vince-fugnitto commented:
@svenefftinge the build still fails due to some leftovers

marcdumais-work commented:
Seems related to this commit:

```
commit 103c95e495f5a0739e70ea6ef14fe2489e1ddfe8
Author: Anton Kosyakov <anton.kosyakov@typefox.io>
Date:   Thu Nov 15 16:24:15 2018 +0000

    [shell] get rid of WidgetTracker
    
    Signed-off-by: Anton Kosyakov <anton.kosyakov@typefox.io>
```

marcdumais-work commented:
build error: 
`[compile] src/browser/frontend-application-module.ts(65,10): error TS2305: Module '"/opt/ericsson-dev/theia/packages/core/src/browser/widgets/index"' has no exported member 'WidgetTracker'.
`

marcdumais-work commented:
Fix: https://github.com/theia-ide/theia/pull/3555
